[FLINK-33099][autoscaler] Introduce the Standalone Autoscaler and support flink cluster

1996fanrui opened the pull request:
## What is the purpose of the change

[FLINK-33099][autoscaler] Introduce the Standalone Autoscaler and support flink cluster

## Brief change log

- Introduce the Standalone Autoscaler
- Implement `LoggableEventHandler` and `RescaleApiScalingRealizer`
- Support flink cluster: `FlinkClusterJobListFetcher`

## Verifying this change

Still developing the test.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes
  - The public API, i.e., is any changes to the `CustomResourceDescriptors`: no
  - Core observer or reconciler logic that is regularly executed: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs in the next PR

gyfora commented:
I think the standalone implementation and the new interfaces / tests etc specific to that should be in a different maven module (`flink-autoscaler-standalone`). That way we wouldn't pollute the core implementation with unnecessary interfaces such as the job list etc.

If we are not sure about the stability we could even put this under the examples for now but that may be an overkill.

What do you think @mxm ?

gyfora commented on flink-autoscaler/src/main/java/org/apache/flink/autoscaler/standalone/JobListFetcher.java:
```diff
@@ -0,0 +1,31 @@
+@Experimental
+public interface JobListFetcher<KEY> {
+
+    List<JobAutoScalerContext<KEY>> fetch(MetricGroup metricGroup) throws Exception;
```
Do we really need this interface in the initial version? Can we simply go with the implementation?

mxm replied:
+1, let's just keep the implementation. There are no other consumers for this interface.

1996fanrui replied:
Of course, I can remove this interface in this PR first, and add it in FLINK-33100.

mxm replied:
I'm also ok with keeping the interface.

gyfora commented on flink-autoscaler/src/main/java/org/apache/flink/autoscaler/standalone/JobListFetcherFactory.java:
```diff
@@ -0,0 +1,30 @@
+@Experimental
+public interface JobListFetcherFactory<KEY> {
+
+    JobListFetcher<KEY> create(ParameterTool parameters, Duration restClientTimeout);
```
Do we really need this factory in the initial version?

mxm replied:
+1 We can construct the implementation directly.

1996fanrui replied:
I can remove this interface in this PR first, and add it in [FLINK-33100](https://issues.apache.org/jira/browse/FLINK-33100).

1996fanrui reviewed (commented):
> I think the standalone implementation and the new interfaces / tests etc specific to that should be in a different maven module (flink-autoscaler-standalone). That way we wouldn't pollute the core implementation with unnecessary interfaces such as the job list etc.
> 
> If we are not sure about the stability we could even put this under the examples for now but that may be an overkill.

Sounds make sense to me. 

Could we just move the standalone package into the `flink-autoscaler-standalone` module and keep `LoggableEventHandler`, `RescaleApiScalingRealizer` and rest body related classes in the `flink-autoscaler` module?

gyfora commented:
The `RescaleApiScalingRealizer` should probably move to the standalone implementation as well. Instead of introducing a LoggableEventHandler, we could simply add default implementation to the interface methods directly.

gyfora reviewed (changes requested):
I think we should remove the new interfaces / factory interfaces. We don't need interfaces and factories for everything. This is a concrete standalone implementation let's keep the code change minimal. We can always add interfaces later if we want to extend this.

gyfora commented on flink-autoscaler/src/main/java/org/apache/flink/autoscaler/standalone/flinkcluster/FlinkClusterJobListFetcher.java:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    @Override
+    public List<JobAutoScalerContext<JobID>> fetch(MetricGroup metricGroup) throws Exception {
```
I don't think `metricGroup` should be part of the method argument. The `FlinkClusterJobListFetcher` could under the hood simply pass the unregistered metric group.

1996fanrui replied:
Good catch! It can be simply, it can be passed to `FlinkClusterJobListFetcher` by the constructor directly.

Also, I don't pass it in this PR, and `new UnregisteredMetricsGroup()` directly when creating the `JobAutoScalerContext`, because we don't support metric reporter for Standalone Autoscaler so far.

mxm reviewed (commented):
Looks good! I would also like to see a dedicated module for these classes. Great work!

1996fanrui commented on flink-autoscaler/src/main/java/org/apache/flink/autoscaler/realizer/RescaleApiScalingRealizer.java:
```diff
@@ -0,0 +1,153 @@
+
+        JobStatus jobStatus = context.getJobStatus();
+        JobID jobID = context.getJobID();
+        if (jobID == null
+                || jobStatus == null
+                || jobStatus.isGloballyTerminalState()
+                || JobStatus.RECONCILING == jobStatus) {
+            LOG.info("Job in terminal or reconciling state cannot be scaled in-place");
+            return;
+        }
```
I try to keep the same logic with `NativeFlinkService#supportsInPlaceScaling`[1]. After thinking again, I think it can be simply: just check for `RUNNING` state.

Because `JobAutoScalerImpl#scale` won't runScalingLogic  if it isn't `RUNNING`.

[1] https://github.com/apache/flink-kubernetes-operator/blob/582c07aeec6b4bbcc68e0a85d31000d89dbd1958/flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/service/NativeFlinkService.java#L260

1996fanrui reviewed (commented):
Hi @gyfora  @mxm , thanks for the quick review.

I have updated all comments except https://github.com/apache/flink-kubernetes-operator/pull/698#discussion_r1377377454. 

And I will try to finish the tests asap.

gyfora commented on flink-autoscaler-standalone/src/main/java/org/apache/flink/autoscaler/standalone/StandaloneAutoscalerEntrypoint.java:
```diff
@@ -0,0 +1,135 @@
+    public static final String FLINK_CLUSTER_PORT = "flinkClusterPort";
+    private static final int DEFAULT_FLINK_CLUSTER_PORT = 8081;
+
+    public static <KEY> void main(String[] args) {
```
If we are adding the fatjar + also include the respective notice file

1996fanrui replied:
Thanks for your reminder!

Samrat002 reviewed (commented):
thank you @1996fanrui , 
overall  +1 for the patch 

I have few queries:
1. can you please help understand how `YarnJobFetcher` can be used in hadoop cluster and deploy flink-yarn ? How exactly it can be used to serve autoscaling ? 

2. will `YarnJobFetcher` be a seperate process or demon in flink-yarn cluster ?

1996fanrui commented:
Hi @gyfora  @mxm , I have addressed all comments, added README, finished all unit tests and ran it locally with fat jar. Please take a look in your free time, thanks~

> If we are adding the fatjar + also include the respective notice file.

Hi @gyfora , I didn't understand the notice file. Could you elaborate on it?


-------------------------------


>  2.  will `YarnJobFetcher` be a seperate process or demon in flink-yarn cluster ?

Hi @Samrat002 , thank you for your review!

I have added a `README.md` to explain some details of `Standalone Autoscaler`, it includes how it works with `JobListFetcher`. `YarnJobListFetcher` is an implemetation of `JobListFetcher`, I think they are similar. you can take a look, thanks~


>     1. can you please help understand how `YarnJobFetcher` can be used in hadoop cluster and deploy flink-yarn ? How exactly it can be used to serve autoscaling ?

As I understand, users need to pass the `yarn resource manger address` to  `YarnJobListFetcher`, `YarnJobListFetcher` can fetch job list by 2 steps:
1 . `YarnJobListFetcher` can fetch all flink clusters from yarn cluster based on rest api. 
2. Step2 is totally same with current `FlinkClusterJobListFetcher`, because `FlinkClusterJobListFetcher` works with single flink cluster.

Please correct me if my understanding is wrong, thanks~

gyfora commented on flink-autoscaler-standalone/README.md:
```diff
@@ -0,0 +1,79 @@
+`Flink Autoscaler Standalone` is an implementation of `Flink Autoscaler`, it runs as 
+a separate java process. It computes the reasonable parallelism of all job vertices 
+by monitoring the metrics, such as: processing rate, busy time, etc. Please see 
+[FLIP-271](https://cwiki.apache.org/confluence/display/FLINK/FLIP-271%3A+Autoscaling) 
```
Here can we link the operator docs autoscaler page instead? Lot of things have changed in the Flip and that page will never be updated.

gyfora commented on flink-autoscaler-standalone/README.md:
```diff
@@ -0,0 +1,79 @@
+for an overview of how autoscaling works.
+
+`Flink Autoscaler Standalone` rescales flink job in-place by rescale api of 
+[FLIP-291](https://cwiki.apache.org/confluence/x/9pRbDg).
```
Here again we should link the docs instead: https://nightlies.apache.org/flink/flink-docs-release-1.18/docs/deployment/elastic_scaling/#externalized-declarative-resource-management

gyfora commented on flink-autoscaler-standalone/README.md:
```diff
@@ -0,0 +1,79 @@
+# Flink Autoscaler Standalone
+
+## What's the autoscaler standalone?
```
We should add a paragraph clarifying that this should only be used in non-kubernetes environments. For Kubernetes the Operator provides a full integration that is preferable.

gyfora commented on flink-autoscaler-standalone/pom.xml:
```diff
@@ -0,0 +1,207 @@
+            <artifactId>flink-runtime</artifactId>
+            <version>${flink.version}</version>
+            <exclusions>
+                <exclusion>
+                    <artifactId>flink-rpc-akka-loader</artifactId>
+                    <groupId>org.apache.flink</groupId>
+                </exclusion>
+                <exclusion>
+                    <artifactId>flink-shaded-zookeeper-3</artifactId>
+                    <groupId>org.apache.flink</groupId>
+                </exclusion>
+                <exclusion>
+                    <artifactId>flink-queryable-state-client-java</artifactId>
+                    <groupId>org.apache.flink</groupId>
+                </exclusion>
+                <exclusion>
+                    <artifactId>snappy-java</artifactId>
+                    <groupId>org.xerial.snappy</groupId>
+                </exclusion>
+                <exclusion>
+                    <artifactId>flink-hadoop-fs</artifactId>
+                    <groupId>org.apache.flink</groupId>
+                </exclusion>
+                <exclusion>
+                    <artifactId>commons-compress</artifactId>
+                    <groupId>org.apache.commons</groupId>
+                </exclusion>
+                <exclusion>
+                    <artifactId>commons-text</artifactId>
+                    <groupId>org.apache.commons</groupId>
+                </exclusion>
+                <exclusion>
```
In this entire pom we have groupId / artifactId , in the reverse order. While syntactically correct this is not how it is generally written. Can you please change them?

gyfora commented on flink-autoscaler-standalone/pom.xml:
```diff
@@ -0,0 +1,207 @@
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-shade-plugin</artifactId>
+                <executions>
+                    <!-- Run shade goal on package phase -->
+                    <execution>
```
As we are bundling dependencies into a father and we are likely going to publish that to maven (right?) .

We need to add a NOTICE file to list the bundled dependencies by license

1996fanrui replied:
Thanks @gyfora  for the quick review! The rest comments are addressed.

> As we are bundling dependencies into a father and we are likely going to publish that to maven (right?) .

During add the fat jar, I have a question: where do users to download this fat jar? In maven repo or dist?

> We need to add a NOTICE file to list the bundled dependencies by license

Sorry, I didn't do it before. Is the NOTICE file created by tools or manually?

gyfora replied:
We could simply publish the fat jar to maven, that would be the simplest (instead of the slim jar)

As for the notice file creation, I don't know any tools, but I think you will not have too many things in it as it's mostly Flink (that do not need to be listed)

1996fanrui replied:
> jar

I have deployed this PR to snapshot repo: https://repository.apache.org/content/repositories/snapshots/org/apache/flink/flink-autoscaler-standalone/1.7-SNAPSHOT/

Run it locally, it works well.

> NOTICE

done

mateczagany reviewed (changes requested):
I only have 2 very minor comments, other than that it looks great.

1996fanrui commented on flink-autoscaler-standalone/src/main/java/org/apache/flink/autoscaler/standalone/StandaloneAutoscalerEntrypoint.java:
```diff
@@ -0,0 +1,121 @@
+
+    public static <KEY, Context extends JobAutoScalerContext<KEY>> void main(String[] args) {
+        var parameters = ParameterTool.fromArgs(args);
+        LOG.info("The standalone autoscaler is started, parameters: {}", parameters);
```
Thanks @mateczagany for your review!

Good catch! I have updated, and tried it locally, it works well.

```
2023-11-03 18:18:54,002 o.a.f.a.s.StandaloneAutoscalerEntrypoint [INFO ] [.] The standalone autoscaler is started, parameters: {flinkClusterPort=8081, flinkClusterHost=localhost}

```

1996fanrui commented on flink-autoscaler-standalone/src/test/java/org/apache/flink/autoscaler/standalone/StandaloneAutoscalerExecutorTest.java:
```diff
@@ -0,0 +1,122 @@
+                        });
+
+        // scaling shouldn't throw exception even if fetch fails
+        autoscalerExecutor.scaling();
```
Technically, I don't know why need this. Because the test will fail if `autoscalerExecutor.scaling();` throw any exception.

From a code design perspective, explicitly declaring `doesNotThrow` may be more clear for code maintenance. So it sounds good to me. I updated.

mxm commented on flink-autoscaler/src/main/java/org/apache/flink/autoscaler/event/AutoScalerEventHandler.java:
```diff
@@ -32,32 +35,45 @@
 @Experimental
-public interface AutoScalerEventHandler<KEY, Context extends JobAutoScalerContext<KEY>> {
-    String SCALING_SUMMARY_ENTRY =
+public class AutoScalerEventHandler<KEY, Context extends JobAutoScalerContext<KEY>> {
```
I think this should remain an interface because this is meant to be implemented by other backends. We already have two implementation of this: Kubernetes, Standalone, and Logging only.

1996fanrui replied:
Both are fine to me, updated the `AutoScalerEventHandler` as an interface and added a `LoggableEventHandler` as the logging  implementation.

mxm commented on flink-autoscaler-standalone/README.md:
```diff
@@ -0,0 +1,83 @@
+
+`Autoscaler` as a generic autoscaler component defines a series of generic interfaces:
+
+- **AutoScalerEventHandler** : Handling autoscaler events, such as: ScalingReport, 
```
This should be an interface, currently a class. Additionally, we can provide a logging implementation.

mxm commented on flink-autoscaler-standalone/README.md:
```diff
@@ -0,0 +1,83 @@
+Updating the `flinkClusterHost` and `flinkClusterPort` based on your flink cluster. 
+In general, the host and port are the same as Flink WebUI.
+
+## Extensibility of autoscaler standalone
+
+`Autoscaler` as a generic autoscaler component defines a series of generic interfaces:
+
+- **AutoScalerEventHandler** : Handling autoscaler events, such as: ScalingReport, 
+  AutoscalerError, etc. It logs events by default.
+- **AutoScalerStateStore** : Storing all state during scaling. `InMemoryAutoScalerStateStore` 
+  is the default implementation, it's based on the Java Heap, so the state will be discarded 
+  after process restarts. We will implement persistent State Store in the future, such as
+  : `JdbcAutoScalerStateStore`.
+- **ScalingRealizer** : Applying scaling actions. `RescaleApiScalingRealizer` is the default
+  implementation, it uses the Rescale API to apply parallelism changes.
+- **JobAutoScalerContext** : Including all details related to the current job.
```
Line 61-73 should be part of the flink-autoscaler README.

1996fanrui replied:
done

mxm reviewed (approved):
Great work! LGTM, just some minor comments.

gyfora commented on flink-autoscaler/src/main/java/org/apache/flink/autoscaler/event/LoggableEventHandler.java:
```diff
@@ -0,0 +1,52 @@
+import java.time.Duration;
+
+/** Autoscaler event handler which logs events. */
+public class LoggableEventHandler<KEY, Context extends JobAutoScalerContext<KEY>>
```
This should be called `LoggingEventHandler`

1996fanrui replied:
Thanks for the review, updated!

gyfora reviewed (approved):
Added one last minor comment, otherwise looks good!

gyfora reviewed (approved):
🚢

1996fanrui commented:
Thanks a lot everyone for the review, merging~
